### DISCUSSION TITLE
How to effectively handle infinite loops?

### DISCUSSION THREAD

**Discussion by Sebight** [berry-lang/berry#430] [Q&A] [answered]
Hello, I have successfully implemented most of Berry support into our firmware and now I am wondering how can I effectively detect infinite loops and stop them.

I saw this macro:
![image](https://github.com/berry-lang/berry/assets/59232422/36f1f700-ec79-42c0-a126-cf997b60e825)

But even after decreasing it to as down as value 2, I still get stuck in an infinite loop. The good thing is that it triggers the ESP watchdog, so one possible solution is clearing the saved Berry code if the reset reason was WDT task. However, this could bring up unwanted behaviour.

Any other ideas or experience, maybe the way you handle it in Tasmota? @s-hadinger 

Thank you!

**Comment by s-hadinger** [comment, marked as answer]
Sure, the heartbeat is managed here: https://github.com/arendst/Tasmota/blob/4fbe3590bea16b3d43fc870bda13807e5d223534/tasmota/tasmota_xdrv_driver/xdrv_52_9_berry.ino#L292-L301

And the observability callback is registered here: `be_set_obs_hook(berry.vm, &BerryObservability);  /* attach observability hook */`

So every 1 million instructions, it has the opportunity to raise an exception and kill the infinite loop.

**Reply by Sebight** [reply]
Perfect, thanks for the resource. I'll fully implement it tomorrow. Just from playing around with `be_raise`, it seems to call ESP32's `abort` function. Is that the expected behaviour, or should it call a catchable C++ exception?

Thank you!

**Reply by s-hadinger** [reply]
Uncaught exceptions result in `abort` indeed. I always call Berry code with `be_pcall()` that reports any exception inside the code, for ex: https://github.com/arendst/Tasmota/blob/4fbe3590bea16b3d43fc870bda13807e5d223534/tasmota/tasmota_xdrv_driver/xdrv_52_9_berry.ino#L154-L160

`be_pcall()` is "protected call", i.e. it will catch any error or exception and not raise to higher levels

**Reply by Sebight** [reply]
Got it, I was trying out a quick example, and I called the `be_raise` before even calling the function using `be_pcall`, but I understand it now.

Thanks a lot for the swift response and plenty of examples.
